jq.lua: Use temp_file strategy for non-Win32 machines

jaredbarranco opened the pull request:
This method correctly handles single quotes and allows for better debugging of the plugin when users have questions. If a method/function errors out the ~/.local/share/nvim/lazy/json-nvim/temp.json or other PLUGIN_HOME dir will have the "in-progress" file.

jaredbarranco committed:
Use temp file for non-windows jq actions

use temp_file for non-windows is_valid

debug is_valid

back to original

debug loggers

more debug loggers

fix debug

log exit_status

remove redundant debugs, add validity check debug

bool to string

debugging

debugging

debugging

debugging
In jq.lua, use temp file for jq actions.

remove debug line

jaredbarranco committed:
Merge pull request #1 from jaredbarranco/feat/heredoc-buffer-read

Use temp file for non-windows jq actions

jaredbarranco commented:
Because of issues related in Issue #2

jaredbarranco commented:
@VPavliashvili I did some preliminary testing with some of the functions I am commonly using. 

I'll be making another PR in the next couple of weeks for some additional improvements if you're open to it.

jaredbarranco committed:
use a temporary directory and epoch/operation for file operations

jaredbarranco committed:
remove apparent non-required code

The init_path logic below seems to account for OS path separators

jaredbarranco committed:
use OS specific temp directory

For OS management of temp files

jaredbarranco committed:
use correct arg.input

VPavliashvili commented:
@jaredbarranco I started to run tests too and found some problems
for example take this json(generated from [online generator](https://json-generator.com/))

```
[
  {
    "_id": "67b3739c332882c1456e076b",
    "index": 0,
    "guid": "9a249e2b-173b-4ab9-ac77-9257d780eedb",
    "isActive": true,
    "balance": "$3,456.75",
    "picture": "http://placehold.it/32x32",
    "age": 28,
    "eyeColor": "green",
    "name": "Olga Hess",
    "gender": "female",
    "company": "REPETWIRE",
    "email": "olgahess@repetwire.com",
    "phone": "+1 (920) 506-2554",
    "address": "918 Fillmore Avenue, Chapin, Indiana, 8526",
    "about": "Proident ea enim sit commodo ullamco minim sint non. Cupidatat reprehenderit pariatur sit in in pariatur reprehenderit consectetur occaecat sit. Consectetur excepteur nisi adipisicing pariatur ullamco labore voluptate elit in do.\r\n",
    "registered": "2016-07-23T05:21:41 -04:00",
    "latitude": -38.870975,
    "longitude": 139.528881,
    "tags": [
      "nisi",
      "aliquip",
      "exercitation",
      "id",
      "velit",
      "eiusmod",
      "proident"
    ],
    "friends": [
      {
        "id": 0,
        "name": "Constance Mcmillan"
      },
      {
        "id": 1,
        "name": "Alexandria Lamb"
      },
      {
        "id": 2,
        "name": "Amparo Barber"
      }
    ],
    "greeting": "Hello, Olga Hess! You have 8 unread messages.",
    "favoriteFruit": "strawberry"
  }
]
```

When I tried to format after your PR I got this result (run :JsonMinifyFile)
```
[  {    "_id": "67b3739c332882c1456e076b",    "index": 0,    "guid": "9a249e2b-173b-4ab9-ac77-9257d780eedb",    "isActive": true,    "balance": "$3,456.75",    "picture": "http://placehold.it/32x32",    "age": 28,    "eyeColor": "green",    "name": "Olga Hess",    "gender": "female",    "company": "REPETWIRE",    "email": "olgahess@repetwire.com",    "phone": "+1 (920) 506-2554",    "address": "918 Fillmore Avenue, Chapin, Indiana, 8526",    "about": "Proident ea enim sit commodo ullamco minim sint non. Cupidatat reprehenderit pariatur sit in in pariatur reprehenderit consectetur occaecat sit. Consectetur excepteur nisi adipisicing pariatur ullamco labore voluptate elit in do.\r\n",    "registered": "2016-07-23T05:21:41 -04:00",    "latitude": -38.870975,    "longitude": 139.528881,    "tags": [      "nisi",      "aliquip",      "exercitation",      "id",      "velit",      "eiusmod",      "proident"    ],    "friends": [      {        "id": 0,        "name": "Constance Mcmillan"      },      {        "id": 1,        "name": "Alexandria Lamb"      },      {        "id": 2,        "name": "Amparo Barber"      }    ],    "greeting": "Hello, Olga Hess! You have 8 unread messages.",    "favoriteFruit": "strawberry"  }]
```

note extra spacing, I think this is a regress.
before your commit it looks like this:
```
[{"_id":"67b3739c332882c1456e076b","index":0,"guid":"9a249e2b-173b-4ab9-ac77-9257d780eedb","isActive":true,"balance":"$3,456.75","picture":"http://placehold.it/32x32","age":28,"eyeColor":"green","name":"Olga Hess","gender":"female","company":"REPETWIRE","email":"olgahess@repetwire.com","phone":"+1 (920) 506-2554","address":"918 Fillmore Avenue, Chapin, Indiana, 8526","about":"Proident ea enim sit commodo ullamco minim sint non. Cupidatat reprehenderit pariatur sit in in pariatur reprehenderit consectetur occaecat sit. Consectetur excepteur nisi adipisicing pariatur ullamco labore voluptate elit in do.\r\n","registered":"2016-07-23T05:21:41 -04:00","latitude":-38.870975,"longitude":139.528881,"tags":["nisi","aliquip","exercitation","id","velit","eiusmod","proident"],"friends":[{"id":0,"name":"Constance Mcmillan"},{"id":1,"name":"Alexandria Lamb"},{"id":2,"name":"Amparo Barber"}],"greeting":"Hello, Olga Hess! You have 8 unread messages.","favoriteFruit":"strawberry"}]
```

please consider finding the cause and fixing it
EDIT: I should have specified that I run this test on Linux

jaredbarranco committed:
fix: use jq -c on get_collapsed

jaredbarranco committed:
Merge pull request #2 from jaredbarranco/feat/heredoc-buffer-read

fix: use jq -c on get_collapsed

jaredbarranco committed:
Merge branch 'feat/temp-file-naming-update' into feat/heredoc-buffer-read

move json-nvim to json_nvim

move lua/json-nvim to lua/json_nvim

not including files that were update for string input testin

add M. functions that accept a string for operations

minify, format, and escape - accept a string

working test script with input/goldens

delete un-used lua/json_nvim/json-nvim.lua

Change name from Token to Node

github actions workflow

jaredbarranco commented:
@VPavliashvili Thanks for catching that regression. I believe I corrected it with [this commit](https://github.com/VPavliashvili/json-nvim/pull/3/commits/a583ac49793f6ea3ce2a8eaa33a26dd007e55111).

I need to do some additional regression testing before I can confidently say it's ready to go. 

On a related note, I'm going to open an issue for discussion around testing framework - something I think would be helpful for making sure regression like this one doesn't happen in the future.

VPavliashvili commented:
@jaredbarranco  Do you think it's better to incorporate unit tests first and then proceed with this PR or other way around?

jaredbarranco commented:
@VPavliashvili Unit testing first would be best so that we can get a baseline expectation. If possible, I think you should be the one to generate those because I can't run ANY of the functions with master from your repo on Mac and everything but the JsonKeysTo* functions on linux.

VPavliashvili commented:
@jaredbarranco  unfortunately I am on a tight schedule right now and you will have to wait for a little while before I will be able to do the work, otherwise I am down to your proposition. 
Also before that, if you provide some small example of some tests, it would be nice and helpful(ofc not required tho)

EDIT: Lets continue unit testing related discussion in #4 and leave this section specific to the PR

jaredbarranco committed:
typo fix in jq.lua comments

jaredbarranco committed:
Merge branch 'master' into feat/golden-testing

jaredbarranco committed:
update workflow

echo runner temp

cache apt, env. namespace incorrect

fix tmpdir declaration

set temp for windows

jaredbarranco committed:
ignore .actual and .expected in test dir

ignore .expected

jaredbarranco committed:
Merge pull request #5 from jaredbarranco/feat/golden-testing

Feat/golden testing

jaredbarranco committed:
update readme for node name update

jaredbarranco committed:
Merge pull request #6 from jaredbarranco/feat/golden-testing

update readme for node name update
